mis_builder/models/mis_report_instance_annotation.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2025 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+import json
+
+from odoo import api, fields, models
+
+
+class MisReportInstanceAnnotation(models.Model):
+    _name = "mis.report.instance.annotation"
+    _description = "Mis Report Instance Annotation"
+
+    period_id = fields.Many2one(
+        comodel_name="mis.report.instance.period",
+        ondelete="cascade",
+        required=True,
+    )
+    kpi_id = fields.Many2one(
+        comodel_name="mis.report.kpi",
+        ondelete="cascade",
+        required=True,
+    )
+    subkpi_id = fields.Many2one(
+        comodel_name="mis.report.subkpi",
+        ondelete="cascade",
+    )
+    note = fields.Char()
+
+    annotation_context = fields.Json(
+        help="""
+        Context used when adding annotation
+        """
+    )
+
+    _sql_constraints = [
+        (
+            "annotation_unique",
+            "unique(period_id, kpi_id,subkpi_id)",
+            "Only one annotation per cell(defined using period,kpi,sub kpi)",
+        ),
+    ]
+
+    def init(self):
+        self.env.cr.execute(
+            """
+                CREATE INDEX IF NOT EXISTS
+                mis_report_instance_annotation_period_id_kpi_id_subkpi_id_idx
+                ON mis_report_instance_annotation(period_id,kpi_id,subkpi_id);
+            """
+        )
+
+    @api.model
+    def get_first_matching_annotation(self, period_id, kpi_id, subkpi_id, instance_id):
+        """
+        Return first annoation
+        matching exactly the period,kpi,subkpi and annotation context
+        """
+        annotations = self.env["mis.report.instance.annotation"].search(
+            [
+                ("period_id", "=", period_id),
+                ("kpi_id", "=", kpi_id),
+                ("subkpi_id", "=", subkpi_id),
+            ],
+        )
+        annotation_context = self._get_annotation_context(instance_id)
+        annotation = fields.first(
+            annotations.filtered(
+                lambda rec: rec.annotation_context == annotation_context
+            )
+        )
+        return annotation
+
+    @api.model
+    def _get_annotation_context(self, instance_id):
+        return json.dumps(
+            self.env["mis.report.instance"]
+            .browse(instance_id)
+            ._get_annotation_context()
+        )
+
+    @api.model
+    def set_annotation(self, period_id, kpi_id, subkpi_id, instance_id, note):
+        annotation = self.get_first_matching_annotation(
+            period_id, kpi_id, subkpi_id, instance_id
+        )
+
+        if annotation:
+            annotation.note = note
+        else:
+            self.env["mis.report.instance.annotation"].create(
+                {
+                    "period_id": period_id,
+                    "kpi_id": kpi_id,
+                    "subkpi_id": subkpi_id,
+                    "note": note,
+                    "annotation_context": self._get_annotation_context(instance_id),
```

Comment:
```suggestion
                    "annotation_context": json.dumps(instance_id._get_annotation_context()),
```

It's simpler.

But is the `json.dumps` necessary? I'd think we can provide the dict to the json field.

Reply:
Indeed the `json.dumps` can be safely removed